src/psyclone/psyir/transformations/acc_update_trans.py

```diff
@@ -314,54 +287,93 @@ def _add_update_directives(self, node_list):
                 # If within if stmt, update device directives can only be moved
-                # out if an update host directive for the same variable - which
-                # always exists as all outputs are inputs - has already been
-                # moved out as well. This happens iff there's no textually
-                # preceding kernel within the if stmt which writes that
-                # variable. Since that variable would appear in loop_sync as
-                # part of the outputs of all textually preceding kernels, we
-                # may simply place an update device directive inside the if
-                # statement for those variables in loop_sync.
+                # out, i.e. executed unconditionally, if an update host
+                # directive for the same variable - which always exists as all
+                # outputs are inputs - has already been moved out as well. This
+                # happens if and only if there's no textually preceding kernel
+                # within the if stmt which writes that variable. Since that
+                # variable would appear in loop_sync as part of the outputs of
+                # any textually preceding kernels, we may simply place an
+                # update device directive inside the if statement for those
+                # variables in loop_sync. loop_sync also includes the outputs
+                # of this host region which are inputs of any textually
+                # preceding kernels but, even though these might not need to
+                # be updated here, i.e. conditionally, they can safely be.
                 if isinstance(sched.parent, Loop) or \
-                   isinstance(sched.parent, IfBlock) and idx == OUT:
-                    self._place_update(sched, update_pos, loop_sync, direction)
+                   isinstance(sched.parent, IfBlock) and acss_type == OUT:
+                    self._place_update(sched, update_pos, loop_sync, acss_type)
                     host_sig.difference_update(loop_sync)
 
                 # This schedule is the body of a routine and, at least until we
                 # can do interprocedural analysis, this is the end of the road.
                 if isinstance(sched, InvokeSchedule):
-                    self._place_update(sched, update_pos, host_sig, direction)
+                    self._place_update(sched, update_pos, host_sig, acss_type)
                     break
 
                 # Move up the code tree to the next deepest ancestor schedule.
-                while True:
+                child, sched = child.parent, sched.parent
+                while not isinstance(sched, Schedule):
                     child, sched = child.parent, sched.parent
-                    if isinstance(sched, Schedule):
-                        break
 
-    def _place_update(self, sched, update_pos, host_sig, direction):
+    def _sync_sig(self, dep_stmts, host_sig, acss_type):
```

Comment:
Please could you make this name more readable, maybe `_get_signatures_to_sync`?

Reply:
Such a loooong name though, it'll force me to break lines.

Reply:
Done. Kinda.

Reply:
It's not as good as my name but it's good enough ;-)